`gpu(::DataLoader)`, take III (#2245)
FluxML/Flux.jl: docs/src/gpu.md

 0.7766742
```

## Transferring Training Data

In order to train the model using the GPU both model and the training data have to be transferred to GPU memory. Moving the data can be done in two different ways:

> 1. Iterating over the batches in a [`DataLoader`](@ref) object transferring each one of the training batches at a time to the GPU. This is recommended for large datasets. Done hand, it might look like this:

Reviewer: "done hand"?

Author: ```suggestion
1. Iterating over the batches in a [`DataLoader`](@ref) object transferring each one of the training batches at a time to the GPU. This is recommended for large datasets. Done by hand, it might look like this:
```

   ```julia
   train_loader = Flux.DataLoader((X, Y), batchsize=64, shuffle=true)
   # ... model definition, optimiser setup
   for epoch in 1:epochs
       for (x_cpu, y_cpu) in train_loader
           x = gpu(x_cpu)
           y = gpu(y_cpu)
           grads = gradient(m -> loss(m, x, y), model)
           Flux.update!(opt_state, model, grads[1])
       end
   end
   ```

>    Rather than write this every time, there is a method of `gpu(::DataLoader)` which does it for you:

Reviewer: ```suggestion
   Rather than write this every time, the method `gpu(::DataLoader)` does it for you:
```

Author: ```suggestion
   Rather than write this out every time, you can just call `gpu(::DataLoader)`:
```

   ```julia
   gpu_train_loader = Flux.DataLoader((X, Y), batchsize=64, shuffle=true) |> gpu
   # ... model definition, optimiser setup
   for epoch in 1:epochs
       for (x, y) in gpu_train_loader
           grads = gradient(m -> loss(m, x, y), model)
           Flux.update!(opt_state, model, grads[1])
       end
   end
   ```
   This is equivalent to `DataLoader(MLUtils.mapobs(gpu, (X, Y)); keywords...)`.

>    Something similar can also be done with [`CUDA.CuIterator`](https://cuda.juliagpu.org/stable/usage/memory/#Batching-iterator), `gpu_train_loader = CUDA.CuIterator(train_loader)`. However, this only works with a limited number of data types: `first(train_loader)` should be a tuple (or `NamedTuple`) of arrays.

Reviewer: Here we could hint at using `mapobs` to transform the dataset into something CUIterator compatible. Could be a short example like
```julia
train_loader = mapobs(preprocess_transform, train_loader)
gpu_train_loader = CUDA.CuIterator(train_loader)
```
 Also, mention when CuIterator should be preferred over gpu?

Author: Is it preferred? This PR takes the line that it's not... it's tricky about what types, and finalize doesn't matter. So it's mentioned here in case people are already using it. 

If finalize does matter, then we should do https://github.com/FluxML/Flux.jl/pull/2240 instead.

2. Transferring all training data to the GPU at once before creating the `DataLoader`. This is usually performed for smaller datasets which are sure to fit in the available GPU memory.
   ```julia
   gpu_train_loader = Flux.DataLoader((X, Y) |> gpu, batchsize = 32)
   # ...
   for epoch in 1:epochs
       for (x, y) in gpu_train_loader
           # ...
   ```

>    Here `(X, Y) |> gpu` applies [`gpu`](@ref) to both arrays -- it recurses into not just tuples, as here, but also whole Flux models.

Reviewer: ```suggestion
   Here `(X, Y) |> gpu` applies [`gpu`](@ref) to both arrays -- it recurses into not just tuples, as here, but also nested data structures.
```

Author: ```suggestion
   Here `(X, Y) |> gpu` applies [`gpu`](@ref) to both arrays, as it recurses into structures.
```
